plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/credentials/sso/SsoAccessTokenProvider.kt

```diff
@@ -443,39 +472,109 @@ class SsoAccessTokenProvider(
+            AwsTelemetry.openCredentials(
+                result = Result.Failed,
+                reason = "Failed to load PKCE client registration from cache",
+                reasonDesc = e.message
+            )
+            throw e
         }
 
     private fun saveClientRegistration(registration: ClientRegistration) {
-        when (registration) {
-            is DeviceAuthorizationClientRegistration -> {
-                cache.saveClientRegistration(dagClientRegistrationCacheKey, registration)
-            }
+        val credentialType: String
+        try {
+            when (registration) {
+                is DeviceAuthorizationClientRegistration -> {
+                    credentialType = DeviceAuthorizationClientRegistration::class.java.name
+                    cache.saveClientRegistration(dagClientRegistrationCacheKey, registration)
+                }
 
-            is PKCEClientRegistration -> {
-                cache.saveClientRegistration(pkceClientRegistrationCacheKey, registration)
+                is PKCEClientRegistration -> {
+                    credentialType = PKCEClientRegistration::class.java.name
+                    cache.saveClientRegistration(pkceClientRegistrationCacheKey, registration)
+                }
             }
+        } catch (e: Exception) {
+            AwsTelemetry.createCredentials(
+                result = Result.Failed,
+                reason = "Failed to save client registration to cache",
+                reasonDesc = e.message
+            )
+            throw e
         }
+        AwsTelemetry.createCredentials(
+            result = Result.Succeeded,
+            reason = "$credentialType successfully written to cache",
+        )
     }
 
     private fun invalidateClientRegistration() {
-        cache.invalidateClientRegistration(dagClientRegistrationCacheKey)
-        cache.invalidateClientRegistration(pkceClientRegistrationCacheKey)
+        try {
+            cache.invalidateClientRegistration(dagClientRegistrationCacheKey)
+            cache.invalidateClientRegistration(pkceClientRegistrationCacheKey)
+        } catch (e: IOException) {
+            AwsTelemetry.modifyCredentials(
+                credentialModification = CredentialModification.Delete,
+                result = Result.Failed,
+                reason = "Failed to invalidate client registration",
+                reasonDesc = e.message,
+                source = "SsoAccessTokenProvider.invalidateClientRegistration"
+            )
+        }
     }
 
-    private fun saveAccessToken(token: AccessToken) {
-        when (token) {
-            is DeviceAuthorizationGrantToken -> {
-                cache.saveAccessToken(dagAccessTokenCacheKey, token)
+    private fun reauthExpiredRegistration(expiredToken: AccessToken): ClientRegistration? {
+        when (expiredToken) {
+            is DeviceAuthorizationGrantToken -> registerDAGClient()
+            is PKCEAuthorizationGrantToken -> registerPkceClient()
+        }
```

Comment:
a token can only be refreshed by the exact client registration that requested the token.

also if we generate a new client registration then we need to invalidate all tokens that use that registration because of the aforementioned reason

Reply:
I see what you mean now about why loading from the disk after this call is unnecessary -- I wanted to return a reauthorized clientRegistration back to the refresh() function but I can see that is possible with just this register call. 

I do not understand exactly what is required for reauthorization when the clientRegistration is expired if generating a new one invalidates all tokens.